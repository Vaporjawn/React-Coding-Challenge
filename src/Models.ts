
export enum Market {
  SoCal = "SoCal",
  NorCal = "NorCal"
}

export const MarketButtons: {text: string, key: string}[] = [
  {text: "SoCal", key: "Southern California"},
  {text: "NorCal", key: "Northern California"},
  {text: "Reset", key: "Reset Interface"}
]

export enum Role {
  AccountExecutive = "Account Executive",
  HeadOfConstruction = "Head Of Construction",
  Concierge = "Concierge"
}

export interface Person {
  name: string
  role: Role
  markets: Market[]
}

export interface activatedPerson {
  Person: Person,
  Selected: string
}

export interface Team {
  persons: Person[]
}

export const People: Person[] = [
  {
    name: "Adam Friedman",
    role: Role.AccountExecutive,
    markets: [Market.NorCal, Market.SoCal]
  },
  {
    name: "Ceci Clark",
    role: Role.HeadOfConstruction,
    markets: [Market.SoCal]
  },
  {
    name: "Courtney Lacy",
    role: Role.Concierge,
    markets: [Market.NorCal, Market.SoCal]
  },
  {
    name: "John Cromwell",
    role: Role.HeadOfConstruction,
    markets: [Market.NorCal]
  },
  {
    name: "Kirsten Pearson",
    role: Role.Concierge,
    markets: [Market.SoCal]
  },
  {
    name: "Michael Carter",
    role: Role.AccountExecutive,
    markets: [Market.NorCal]
  },
]
