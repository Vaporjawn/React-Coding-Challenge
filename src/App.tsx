import React from "react";
import "./App.css";
import ListSelect from "./components/listSelect";

const App = () => {
  return (
    <div>
      <h1 className="centered">Team Selector</h1>
      <ListSelect/>
    </div>
  );
};

export default App;