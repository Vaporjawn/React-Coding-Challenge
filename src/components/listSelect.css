.isSelected{
    background-color: green;
    font-weight: bold;
}
.pictureText{
    text-align: center;
    border-style: double;
    margin-top: 2%;
    margin-left: auto;
    margin-right: auto;
    width: 300px;

}
.flex{
    display: flex;
}
.buttonBar{
    margin-left: auto;
    margin-right: auto;
    text-align: center;
}
.button{
    padding: 1%;
    font-size: 25px;
    margin: 1%;
}
.visible{
    display: flex;
    pointer-events: none;
}
.invisible{
    display: none;
}
.centered{
    text-align: center;
}
.error{
    color: red;
}
.TeamPicture{
    -webkit-user-drag: none;
}