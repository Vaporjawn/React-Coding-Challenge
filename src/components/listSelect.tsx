import React, { useEffect, useState } from "react";
import { People, MarketButtons, Person, Role, activatedPerson} from "../Models";
import AdamFriedman from '../team_images/adam-friedman.jpg';
import CeciClarke from '../team_images/ceci-clarke.jpg';
import CourtneyLacy from '../team_images/courtney-lacy.jpg';
import JohnCromwell from '../team_images/jon-cromwell.jpg';
import KirstenPearson from '../team_images/kirsten-pearson.jpg';
import MichaelCarter from '../team_images/michael-carter.jpg';
import './listSelect.css';

export let dynamicPeople = People;
export let potentialTeam: activatedPerson[] = [];
export let market: string[] = [];
export let teamVisibility: string = 'invisible';
export let resultText: {
    text: string,
    visibility: string,
} = {
    text: '',
    visibility: 'invisible'
}

const getTeamImage = (name: string) => {
    switch(name){
        case 'Adam Friedman':{
            return <img className='TeamPicture' src={AdamFriedman} alt={name}/>
        }
        case 'Ceci Clark':{
            return <img className='TeamPicture' src={CeciClarke} alt={name}/>
        }
        case 'Courtney Lacy':{
            return <img className='TeamPicture' src={CourtneyLacy} alt={name}/>
        }
        case 'John Cromwell':{
            return <img className='TeamPicture' src={JohnCromwell} alt={name}/>
        }
        case 'Kirsten Pearson':{
            return <img className='TeamPicture' src={KirstenPearson} alt={name}/>
        }
        case 'Michael Carter':
            return <img className='TeamPicture' src={MichaelCarter} alt={name}/>
        }
    }

const filterPeople = (marketName: string) => {
    if(marketName === 'Reset'){dynamicPeople = People; potentialTeam = []; resultText.visibility = 'invisible';}
    else{
        dynamicPeople = People;
        dynamicPeople = dynamicPeople.filter(c => c.markets.find(d => d === marketName));
        potentialTeam = potentialTeam.filter(p => p.Person.markets.find(q => q === marketName));
        market = [marketName];
    }
}

const returnArray = (selected: Person) => {
    const chosenElement = potentialTeam.find(c => c.Person === selected);
    if(chosenElement !== undefined){
        return chosenElement.Selected;
    }
}

const marketCheck = (incomingPerson: Person) => {
    if(potentialTeam.length === 0){
        market = incomingPerson.markets;
    }
    if(potentialTeam.length === 1 && potentialTeam[0].Person.markets.length > market.length){
        market = potentialTeam[0].Person.markets;
    }
    if(market.length === 0 && incomingPerson.markets.filter(e => market.indexOf(e))){
        market = incomingPerson.markets;
    }
    if(market.length !== 0 && incomingPerson.markets.filter(e => market.indexOf(e))){
        const staticMarket = market;
        for(let i=0; i < market.length; i++){

            if(incomingPerson.markets.find(m => m === staticMarket[i])){
                if(incomingPerson.markets.length < market.length){
                    market = incomingPerson.markets;
                }
                return true;
            }
        }
        return false;
    }
}

const isSelected = (currentPerson: Person) => {
    if(!potentialTeam.find(p => p.Person === currentPerson) && potentialTeam.length < 3 && !potentialTeam.find(p => p.Person.role === currentPerson.role)){
        if(potentialTeam.find(p => p.Person.markets.find(m => m === currentPerson.markets.find(f => f === m))) || potentialTeam.find(p => p.Person.markets.find(m => m === currentPerson.markets.find(f => f === m))) === undefined){
            let newAddition: activatedPerson = {
                Person: currentPerson,
                Selected: 'isSelected'
            };
            if(marketCheck(currentPerson)){
                potentialTeam.push(newAddition);
            }
        }
    }else{
        potentialTeam = potentialTeam.filter(p => p.Person !== currentPerson);
        if(potentialTeam.length === 0){
            market = []
        }else{
            market = potentialTeam[0].Person.markets;
        }
    }
    if(potentialTeam.length < 3){teamVisibility = 'invisible'; resultText.visibility = 'invisible';};
}

const makeTeam = () => {
    if(potentialTeam.find(x => x.Person.role === Role.AccountExecutive) 
    && potentialTeam.find(x => x.Person.role === Role.Concierge)
    && potentialTeam.find(x => x.Person.role === Role.HeadOfConstruction)){

        teamVisibility = 'flex visible';
        resultText.text = 'Created Team'
        resultText.visibility = 'block'
    }else{
        teamVisibility = 'invisible'
        resultText.text = 'Error: Team Member(s) are missing from team';
        resultText.visibility = 'block error';
    }
}

const ListSelect = () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [time, setTime] = useState(Date.now());

    useEffect(() => {
        const interval = setInterval(() => setTime(Date.now()), 10);
        return () => {
            clearInterval(interval);
        };
    }, []);
        const createdTeam = potentialTeam.map(item =>
            <div className="pictureText">
            <div className={returnArray(item.Person)} onClick={() => isSelected(item.Person)}>
                {getTeamImage(item.Person.name)}<br/>
                {item.Person.name} <br/>
                {item.Person.role} <br/>
                {item.Person.markets.map(c => c + '\n')} <br/><br/>
            </div>
        </div>
            )
        const renderedOutput = dynamicPeople.map(item => 
        <div className="pictureText" key={item.name}>
            <div className={returnArray(item)} onClick={() => isSelected(item)}>
                {getTeamImage(item.name)}<br/>
                {item.name} <br/>
                {item.role} <br/>
                {item.markets.map(c => c + '\n')} <br/><br/>
            </div>
        </div>)
        const filterButtons = MarketButtons.map(item => <button key={item.key} className="button" onClick={() => filterPeople(item.text)}> {item.text} </button>)

    return (
        <div>
            <div className="buttonBar">
                {filterButtons}
                <button className="button" onClick={() => makeTeam()} >Make Team</button>
            </div>
            <div className="flex">
                {renderedOutput}
            </div>
            <div className={resultText.visibility}>
                <h2 className="centered">{resultText.text}</h2>
            </div>
            <div className={teamVisibility}>
                {createdTeam}
            </div>
        </div>
    );
}

export default ListSelect;