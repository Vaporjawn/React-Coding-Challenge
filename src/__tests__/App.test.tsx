import React from "react";
import ReactDOM from "react-dom";
import App from "../App";
import { dynamicPeople, market, potentialTeam, resultText, teamVisibility } from "../components/listSelect";
import { render } from "../test_utils/testRender";
import { People, Market, MarketButtons, Role} from '../Models';


describe("App", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders and matches snapshot", async () => {    
    const c = await render(<App />);
    expect(c.baseElement).toMatchSnapshot();
  });
});

describe("listSelect", () => {
  it("renders without crashing", () => {

  });
});

describe("potentialTeam", () => {
  it("has a max of 3", () => {
    expect(potentialTeam.length).toBeLessThanOrEqual(3);
  });

  it("Starts as empty array", () => {
    expect(potentialTeam).toEqual([]);
  });
});

describe("market", () => {
  it("does not exceed the max number of markets", () => {
    expect(market.length).toBeLessThanOrEqual(2);
  });

  it("Starts as empty array", () => {
    expect(market).toEqual([]);
  });
});

describe("dynamicPeople", () => {
  it("does not become bigger than People", () => {
    expect(dynamicPeople.length).toBeLessThanOrEqual(People.length);
  });
});

describe("teamVisibility", () => {
  it("starts off as invisible", () => {
    expect(teamVisibility).toBe('invisible');
  });
});

describe("resultText", () => {
  it("starts off as invisible", () => {
    expect(resultText).toEqual({
      text: '',
      visibility: 'invisible'
    });
  });
});

describe("Market", () => {
  it("codes are equal to their value", () => {
    expect(Market.NorCal).toBe('NorCal');
    expect(Market.SoCal).toBe('SoCal');
  });
});

describe("Market Buttons", () => {
  it("text contains the correct values", () => {
    expect(MarketButtons[0].text).toBe('SoCal');
    expect(MarketButtons[1].text).toBe('NorCal');
    expect(MarketButtons[2].text).toBe('Reset');
  });

  it("keys contain the correct values", () => {
    expect(MarketButtons[0].key).toBe('Southern California');
    expect(MarketButtons[1].key).toBe('Northern California');
    expect(MarketButtons[2].key).toBe('Reset Interface');
  });

  it("has the correct amount of buttons", () => {
    expect(MarketButtons.length).toEqual(3);
  });
});

describe("Role", () => {
  it("the required roles are available", () => {
    expect(Role.AccountExecutive).toBe('Account Executive');
    expect(Role.HeadOfConstruction).toBe('Head Of Construction');
    expect(Role.Concierge).toBe('Concierge');
  });
});

describe("People", () => {
  it("People never change", () => {
    expect(People.length).toEqual(6);
  });
  
  it("Contains the original members", () => {
    expect(People[0].name).toEqual('Adam Friedman');
    expect(People[1].name).toEqual('Ceci Clark');
    expect(People[2].name).toEqual('Courtney Lacy');
    expect(People[3].name).toEqual('John Cromwell');
    expect(People[4].name).toEqual('Kirsten Pearson');
    expect(People[5].name).toEqual('Michael Carter');
  });
});

