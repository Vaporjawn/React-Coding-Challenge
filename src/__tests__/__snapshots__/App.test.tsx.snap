// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders and matches snapshot 1`] = `
<body>
  <div>
    <div>
      <h1
        class="centered"
      >
        Team Selector
      </h1>
      <div>
        <div
          class="buttonBar"
        >
          <button
            class="button"
          >
             
            SoCal
             
          </button>
          <button
            class="button"
          >
             
            NorCal
             
          </button>
          <button
            class="button"
          >
             
            Reset
             
          </button>
          <button
            class="button"
          >
            Make Team
          </button>
        </div>
        <div
          class="flex"
        >
          <div
            class="pictureText"
          >
            <div>
              <img
                alt="Adam Friedman"
                class="TeamPicture"
                src="adam-friedman.jpg"
              />
              <br />
              Adam Friedman
               
              <br />
              Account Executive
               
              <br />
              NorCal

              SoCal

               
              <br />
              <br />
            </div>
          </div>
          <div
            class="pictureText"
          >
            <div>
              <img
                alt="Ceci Clark"
                class="TeamPicture"
                src="ceci-clarke.jpg"
              />
              <br />
              Ceci Clark
               
              <br />
              Head Of Construction
               
              <br />
              SoCal

               
              <br />
              <br />
            </div>
          </div>
          <div
            class="pictureText"
          >
            <div>
              <img
                alt="Courtney Lacy"
                class="TeamPicture"
                src="courtney-lacy.jpg"
              />
              <br />
              Courtney Lacy
               
              <br />
              Concierge
               
              <br />
              NorCal

              SoCal

               
              <br />
              <br />
            </div>
          </div>
          <div
            class="pictureText"
          >
            <div>
              <img
                alt="John Cromwell"
                class="TeamPicture"
                src="jon-cromwell.jpg"
              />
              <br />
              John Cromwell
               
              <br />
              Head Of Construction
               
              <br />
              NorCal

               
              <br />
              <br />
            </div>
          </div>
          <div
            class="pictureText"
          >
            <div>
              <img
                alt="Kirsten Pearson"
                class="TeamPicture"
                src="kirsten-pearson.jpg"
              />
              <br />
              Kirsten Pearson
               
              <br />
              Concierge
               
              <br />
              SoCal

               
              <br />
              <br />
            </div>
          </div>
          <div
            class="pictureText"
          >
            <div>
              <img
                alt="Michael Carter"
                class="TeamPicture"
                src="michael-carter.jpg"
              />
              <br />
              Michael Carter
               
              <br />
              Account Executive
               
              <br />
              NorCal

               
              <br />
              <br />
            </div>
          </div>
        </div>
        <div
          class="invisible"
        >
          <h2
            class="centered"
          />
        </div>
        <div
          class="invisible"
        />
      </div>
    </div>
  </div>
</body>
`;
